### DISCUSSION TITLE
model.Context is not reusable?

### DISCUSSION THREAD

**Discussion by mdmcconnell** [pdfcpu/pdfcpu#995] [Q&A]
Sorry to be posting lots of questions these days!  
I have a use case where I generate numbers of PDF files, and combine (append) them into a multiple output files.  It might be convenient to do this in the program without writing, reading and removing temporary files.  So I tried this:
```
api.Write(ctxDest, f, pdfConf)
api.Write(ctxDest, f2, pdfConf)
```
But I find the second file written is invalid.  Something must have changed about ctxDest (or pdfConf).  Is it possible to "reset" the Context for a second writing?

**Comment by hhrutter** [comment]
I would use an `io.MultiWriter` if I really don't want to delegate this to the OS.

**Reply by mdmcconnell** [reply]
Thank you very much, that is helpful!  The case is more complex however - I fill several PDF forms, then choose which ones (some duplicates) to output into a couple of combined files.  Since I have representations of the files already in memory, it seems elegant to continue to manipulate them in memory.  I am not being successful at that, so I guess I will have to go through the file system.  Perhaps the Write process updates the Context in various ways.

One other option that could help - is there a way to duplicate a Context?

**Reply by mdmcconnell** [reply]
Hmm, api.MergeCreateFile is not able to deal with this case either, using individual files of filled forms.  After the first filled form, form field data is corrupted.  Looks like a bug, possibly a regression in recent commits.  I will try to put together a simplified example.

**Reply by mdmcconnell** [reply]
A few further notes on this: pdfpcu CLI and qpdf have the same issue trying to merge the files, which suggests it is something unexpectd about the input files rather than the merge process.  I tried qpdf with "flatten-annotations", which complained about missing appearance streams.  qpdf was not able to generate appearance streams correctly.  BUT - I could open with Acrobat, save, and then merge.  So it looks like the lack of appearance streams in my filled form is the issue.

**Comment by hhrutter** [comment]
I cannot recommend merging PDF forms!
You have to understand that one PDF file constitutes one form.
It sounds simple but when merging several PDFs with formfields, all formfields have to be manipulated in order to ensure uniqueness in the final PDF form which is the result of the merge process.

**Reply by mdmcconnell** [reply]
So far I have been more concerned about the visual presentation of the documents, and also have tracked IDs for uniqueness.  But at some point duplication will be an issue to solve.  Does the standard allow for multiple forms in a PDF document?  json structure of PDFs implies that there could be multiple forms, which could conceivably allow non-unique field IDs, and a robust way to merge documents containing forms.

**Comment by hhrutter** [comment]
As per the PDF specification a PDF file may contains zero or one Acrobat form.

Ofcourse when merging you have multiple issues...
What's your strategy when your input files are inconstent about form field appearance streams?
What about multiple form fonts from different forms?

I can only iterate that merging forms can only work in a closed environment where you do have control
about how your input forms are created, anything else ends up in a mess.

**Reply by mdmcconnell** [reply]
I know very little about the PDF file specification at the moment, so it may be a long while before I can answer these questions.  So far, I just have the idea that the specification contemplates the possibility, and I do hope to make contributions some day.  At the moment, the biggest advance for me would be working appearance streams for "comb" fields!

**Comment by mdmcconnell** [comment]
I noticed today that multifill manages to fill several of the same form into a single pdf file.  That must be very similar to merging multiple forms.  I will have to study that code.

**Comment by hhrutter** [comment]
*Multifill* ddresses the use case where a single form is used to generate multiple filled PDF forms based on data from a database, such as creating a series of letters.

**Reply by mdmcconnell** [reply]
Yes, I see!  I was considering the note that "[Optionally this command can merge the output PDFs together.](https://pdfcpu.io/form/form_multifill)", and that the operation of merging many of the same form is substantially the same as merging different forms.  But I haven't investigated as far as even testing this.  

Since I am mainly concerned with visual presentation of the filled forms, for now I flatten annotations into text and then merge PDFs.  I have to execute ghostscript in an external process to flatten, which is not my favourite solution...

**Comment by hhrutter** [comment]
Closing this as answered.

model.Context is not intended to be reused for multiple independent writes. A write operation may update context/output state, so callers should create a fresh context for each independent output.

The later part of this discussion moved into merging filled forms. That remains a difficult area in general: a PDF has at most one AcroForm, and merging form PDFs requires reconciling fields, names, resources, appearances, fonts, and possibly missing appearance streams. For reliable results, flattening before merge is usually the safer approach unless all input forms are generated in a tightly controlled workflow.

multifill covers a different case: generating multiple filled instances from one form template, optionally merging the generated outputs.

Closing for now. Please open a new issue or discussion with a minimal reproducible example if there is still a concrete bug in the current release.
